Dynamic reconfiguration of Server

amitvc opened the pull request:
This pull request is to introduce a feature to dynamically add/remove services/routes etc from a Server. Please see issue - https://github.com/line/armeria/issues/3041 for details.

amitvc commented on core/src/main/java/com/linecorp/armeria/server/Server.java:
```diff
@@ -397,6 +397,12 @@ public String toString() {
+    public void reconfigure(ReconfigurableServer rs) {
+        final ServerBuilder sb = rs.reconfigure(Server.builder());
+        // Here we build our new configuration object. Now time to replace the old config with the new.
+        final ServerConfig config = sb.buildServerConfig(this.config());
```
Here I have built the new ServerConfig. The serverConfig in Server class is final so really cannot replace it. Are we ok to change that ?

trustin replied:
Yeah, I think so.

amitvc commented on core/src/test/java/com/linecorp/armeria/client/TestReconfigurableServer.java:
```diff
@@ -0,0 +1,65 @@
+import com.linecorp.armeria.server.Server;
+import com.linecorp.armeria.server.ServerBuilder;
+
+class TestReconfigurableServer {
```
This is a temporary test code. I will remove it and start adding unit tests when I get the feature more ironed out.

ikhoon replied:
Gentle ping. Any plan to update this? 😉

trustin commented on core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java:
```diff
@@ -1186,6 +1186,14 @@ public VirtualHostBuilder defaultVirtualHost() {
+     * Returns the {@link VirtualHostBuilder} for building the default
+     * <a href="https://en.wikipedia.org/wiki/Virtual_hosting#Name-based">name-based virtual host</a>.
+     */
+    public List<VirtualHostBuilder> defaultVirtualHostBuilders() {
```
Could we hide this method from the public API. Our builder classes don't usually expose getter methods. Is it possible to make it package-private?

amitvc replied:
Makes sense. Will do that.

trustin commented on core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java:
```diff
@@ -1662,4 +1670,35 @@ public String toString() {
                 enableServerHeader, enableDateHeader);
     }
+
+    public final ServerConfig buildServerConfig(ServerConfig existingConfig) {
```
Ditto - This method should not be public but package-private.

amitvc replied:
Yup.

trustin reviewed (commented):
Thank you, @amitvc! Looks like a good start. Looking forward to more updates!

minwoox commented on core/src/main/java/com/linecorp/armeria/server/ReconfigurableServer.java:
```diff
@@ -0,0 +1,6 @@
+
+@FunctionalInterface
+public interface ReconfigurableServer {
+    ServerBuilder reconfigure(ServerBuilder sb);
```
```suggestion
    void reconfigure(ServerBuilder sb);
```

minwoox reviewed (commented):
Looks good. 😄

amitvc commented on core/src/test/java/com/linecorp/armeria/client/TestReconfigurableServer.java:
```diff
@@ -16,50 +16,110 @@
-    }
     @Test
-    public void test() throws InterruptedException, ExecutionException {
+    public void test_reconfiguration_of_server() throws Exception {
```
I think we will need more unit tests and more different combinations of ServerConfig. Maybe add unit tests with ServerBuilder.routes or ServerBuilder.annotatedServices etc?

amitvc commented:
@trustin I have pushed more changes and fix checkstyle issues as well. The build runs successfully on my local but fails on a random UT here [test failure](https://ci.appveyor.com/project/line/armeria/builds/38013355/job/v6b0kl5i18wsg6xd)

minwoox commented on core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java:
```diff
@@ -1662,4 +1662,34 @@ public String toString() {
+        final Mapping<String, SslContext> sslContexts;
+        final SslContext defaultSslContext = findDefaultSslContext(defaultVirtualHost, virtualHosts);
+
+        return new ServerConfig(
```
We need to check if `pingIntervalMillis`, `maxConnectionAgeMillis` and `idleTimeoutMillis` need to be adjusted.
Could you extract the same logic from https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java#L1502-L1576
and apply that for this method as well? 😄

amitvc replied:
Done

minwoox commented on core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java:
```diff
@@ -1662,4 +1662,34 @@ public String toString() {
                 enableServerHeader, enableDateHeader);
     }
+
+    ServerConfig buildServerConfig(ServerConfig existingConfig) {
```
Could you put this method on top of `toString()`?
We usually put `toString()` at the end. 😄

amitvc replied:
Done.

minwoox reviewed (commented):
Thanks, left some comments. 😄

amitvc commented on core/src/main/java/com/linecorp/armeria/server/Server.java:
```diff
@@ -470,9 +483,11 @@ private ChannelFuture doStart(ServerPort port) {
-                                                              gracefulShutdownSupport));
-
+            pipelineConfigurator = new HttpServerPipelineConfigurator(
+                                                                    config,
```
Done.

amitvc commented on core/src/main/java/com/linecorp/armeria/server/Server.java:
```diff
@@ -397,6 +398,18 @@ public String toString() {
+     * @param rs ReconfigurableServer.
+     */
+    public void reconfigure(ReconfigurableServer rs) {
+        final ServerBuilder sb = Server.builder();
```
Done.

amitvc commented on core/src/main/java/com/linecorp/armeria/server/Server.java:
```diff
@@ -397,6 +398,18 @@ public String toString() {
+        final ServerBuilder sb = Server.builder();
+        rs.reconfigure(sb);
+        final ServerConfig newConfiguration = sb.buildServerConfig(this.config());
+        config = newConfiguration;
```
Done.

amitvc commented:
@trustin @minwoox addressed the cr comments.

minwoox commented on core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java:
```diff
@@ -1649,6 +1649,50 @@ private static SslContext findDefaultSslContext(VirtualHost defaultVirtualHost,
+                        .map(vhb -> vhb.build(virtualHostTemplate))
+                        .collect(toImmutableList());
+        // Pre-populate the domain name mapping for later matching.
+        final Mapping<String, SslContext> sslContexts;
```
Because we rebuild the `virtualHosts`, we should create the `sslContexts` again.
How about:
- Adding `sslContexts` into the `ServerConfig`
- Extracting the same logic from `buildServerConfig(ServerConfig existingConfig)` and `build()`?
```java
ServerConfig buildServerConfig(ServerConfig existingConfig) {
    return buildServerConfig(existingConfig.ports());
}

Server build() {
    return new Server(buildServerConfig(ports));
}

ServerConfig buildServerConfig(List<ServerPort> ports) {
    // Setting the parameters...

    return new ServerConfig(...);
}

```

amitvc replied:
Good suggestion. I made the change.

amitvc commented on core/src/main/java/com/linecorp/armeria/server/ReconfigurableServer.java:
```diff
@@ -0,0 +1,28 @@
+public interface ReconfigurableServer {
+    /**
+     * Reconfigures the server using the specified {@link ServerBuilder}.
+     * @param sb ServerBuilder.
```
If I remove the java doc check style complains about it.

minwoox replied:
Hm, that's weird because we don't have to have that.
https://github.com/line/armeria/blob/master/dropwizard2/src/main/java/com/linecorp/armeria/dropwizard/ArmeriaServerConfigurator.java#L29
Do you use another settings?

amitvc replied:
I misunderstood which line you wanted me to delete. I removed the line and checkstyle passes now.

minwoox commented on core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java:
```diff
@@ -1538,14 +1603,15 @@ public Server build() {
-                        "usually by adding a build dependency on the " +
-                        "io.netty:netty-tcnative-boringssl-static artifact or enable Jetty ALPN as described " +
-                        "at https://www.eclipse.org/jetty/documentation/9.4.x/alpn-chapter.html");
+                                "detected. To use TLS with Armeria, you must either use Java 9+, enable " +
```
Could you fix the indentation? 😉

minwoox commented on core/src/test/java/com/linecorp/armeria/client/TestReconfigurableServer.java:
```diff
@@ -0,0 +1,123 @@
+
+        final CompletableFuture<Void> future = server.start();
+        Thread.sleep(1000);
+        try (CloseableHttpClient hc = HttpClients.createMinimal()) {
```
How about using Armeria `WebClient`?
We will remove the Apache client from the test code eventually.

amitvc replied:
Yes will do.

minwoox reviewed (commented):
I think we are almost done. 😄

minwoox commented on core/src/test/java/com/linecorp/armeria/client/TestReconfigurableServer.java:
```diff
@@ -0,0 +1,129 @@
+    @Test
+    public void test_reconfiguration_of_server() throws Exception {
+        final ServerBuilder sb = Server.builder();
+        sb.http(8080);
```
We shouldn't use this specific port because this will fail with `Address already in use` exception when we execute multiple tests concurrently.

amitvc replied:
You are right. I saw this error happen on my local when I ran all unit tests. I will change it.

minwoox commented on core/src/test/java/com/linecorp/armeria/client/TestReconfigurableServer.java:
```diff
@@ -0,0 +1,129 @@
+        final CompletableFuture<Void> future = server.start();
+        Thread.sleep(1000);
+
+        try (ClientFactory factory =
+                     ClientFactory.builder().build()) {
+            final WebClient client = WebClient.builder()
+                    .factory(factory)
+                    .build();
+            final AggregatedHttpResponse response = client.get("http://localhost:8080/test1").aggregate().get();
+            assertThat(response.status()).isEqualTo(HttpStatus.OK);
+            assertThat(response.contentUtf8()).isEqualTo("Hello, world!");
+        }
```
```suggestion
        final AggregatedHttpResponse response =
                WebClient.of("http://127.0.0.1:" + server.activeLocalPort()).aggregate().join();
        assertThat(response.status()).isEqualTo(HttpStatus.OK);
        assertThat(response.contentUtf8()).isEqualTo("Hello, world!");
```

minwoox commented on core/src/test/java/com/linecorp/armeria/client/TestReconfigurableServer.java:
```diff
@@ -0,0 +1,129 @@
+
+        try (ClientFactory factory =
+                     ClientFactory.builder().build()) {
+            final WebClient client = WebClient.builder()
```
Let's just use the client that is created above.

minwoox commented on core/src/test/java/com/linecorp/armeria/client/TestReconfigurableServer.java:
```diff
@@ -0,0 +1,129 @@
+
+            assertThat(response4.status()).isEqualTo(HttpStatus.NOT_FOUND);
+        }
+        future.join();
```
```suggestion
        server.stop().join();
```

minwoox reviewed (commented):
Left few more nits. 😄

trustin commented:
I think switching other properties like ports and TLS properties are probably too much of work for this PR. Shall we focus on re-binding services only? For example, we could enforce a constraint like rejecting any changes unsupported in this PR.

@amitvc Please feel free to take your time and let me know if it's ready for another round of review. We all know this is a challenging topic and it deserves careful thoughts and execution. :smile:

amitvc commented:
@trustin I agree. I think I am still trying to figure out a clean way to track and close channels that were created prior to server.reconfigure(which configures new routes and services).

ikhoon commented on core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java:
```diff
@@ -1499,6 +1499,73 @@ public ServerBuilder annotatedServiceExtensions(
      * Returns a newly-created {@link Server} based on the configuration properties set so far.
      */
     public Server build() {
+        final Server server = new Server(buildServerConfig(this.ports));
```
nit:
```suggestion
        final Server server = new Server(buildServerConfig(ports));
```

amitvc replied:
Done.

ikhoon commented:
> Shall we focus on re-binding services only? For example, we could enforce a constraint like rejecting any changes unsupported in this PR.

do we need `ReconfigurableServerConfigSetters` to mark reconfigurable configs?
```java
interface ReconfigurableServerConfigSetters {
    ReconfigurableServerConfigSetters service(String pathPattern, HttpService service);
    // List up only reconfigurable APIs
    ...
}

class ServerBuilder implements ReconfigurableServerConfigSetters {
    ...
}

class Server {
    public void reconfigure(Consumer<? super ReconfigurableServerConfigSetters> reconfigurator) {
        final ServerBuilder sb = builder();
        reconfigurator.accept(sb);
        ...
    }
}
```

amitvc commented on core/src/main/java/com/linecorp/armeria/server/ReconfigurableServer.java:
```diff
@@ -0,0 +1,27 @@
+/**
+ * A Server that allows ability to reconfigure its service endpoints without having to restart the server.
+ */
+@FunctionalInterface
```
Yes.

amitvc commented on core/src/main/java/com/linecorp/armeria/server/ReconfigurableServer.java:
```diff
@@ -0,0 +1,27 @@
+ * A Server that allows ability to reconfigure its service endpoints without having to restart the server.
+ */
+@FunctionalInterface
+public interface ReconfigurableServer {
```
Awesome. Will replace it. Thanks for pointing this out.

amitvc commented on core/src/main/java/com/linecorp/armeria/server/Server.java:
```diff
@@ -397,6 +398,17 @@ public String toString() {
+     * Reconfigure Server configuration.
+     * @param rs ReconfigurableServer.
+     */
+    public void reconfigure(ReconfigurableServer rs) {
```
will do.

amitvc commented on core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java:
```diff
@@ -106,7 +106,7 @@
     };
 
-    private final ServerConfig config;
+    private ServerConfig config;
```
Good point.

amitvc commented on core/src/main/java/com/linecorp/armeria/server/Server.java:
```diff
@@ -397,6 +398,17 @@ public String toString() {
+        final ServerBuilder sb = builder();
+        rs.reconfigure(sb);
+        config = sb.buildServerConfig(config());
+        pipelineConfigurator.updateConfig(config);
```
Good catch on this. I will guard against this.

amitvc commented:
@trustin @minwoox @ikhoon I have addressed code review comments and added mechanism to notify stale httpServerHandlers about new serverconfig. The thing now left to do is HTTP/2 goaway graceful shutdown' for http2 connections.

amitvc commented on core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java:
```diff
@@ -684,4 +684,13 @@ private RequestId nextRequestId() {
+    protected void performConfigCheck(ServerConfig config) {
+        // This is temporary hack to switch original config with new config for older HttpServerHandlers.
+        // TODO - Figure out a way to close the open channel so these handlers get garbage collected and
+        // not serve any traffic.
```
Also need to figure out a way to remove the subscribers that are stale and no longer needed to be kept around.

amitvc replied:
The above comment is no longer relevant.

amitvc commented:
@trustin @ikhoon @minwoox I have pushed more changes.

amitvc commented:
> @trustin @minwoox @ikhoon I have addressed code review comments and added mechanism to notify stale httpServerHandlers about new serverconfig. The thing now left to do is HTTP/2 goaway graceful shutdown' for http2 connections.

I spoke with Trustin about what should happen to the original HttpServerHandlers which were spawned before the user submits a new ServerConfig. We both agree that switching the old ServerConfig with the new ServerConfig should be enough. We don't need to explicitly close the channel. I discovered a memory leak in my ServerConfigHolder in terms of leaving the callback Consumer<ServerConfig> for the old/inactive HttpServerHandler. I have handled this memory leak by removing callack once the channel is inactive.

minwoox commented on core/src/main/java/com/linecorp/armeria/server/ServerConfigHolder.java:
```diff
@@ -0,0 +1,40 @@
+
+import com.linecorp.armeria.common.util.AbstractListenable;
+
+public final class ServerConfigHolder extends AbstractListenable<ServerConfig> {
```
Do we need this holder even though we don't close the previously established connections?

amitvc replied:
I think we will still need to notify all HttpServerHandlers about ServerConfig changes. This AbstractListenable implementation helps with that. If I don't do this I will have to keep a list of all HttpServerHandler inside HttpServerPipelineConfigurator

trustin commented on core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java:
```diff
@@ -91,6 +92,7 @@
     private static final String MSG_INVALID_REQUEST_PATH = HttpStatus.BAD_REQUEST + "\nInvalid request path";
 
     private static final HttpData DATA_INVALID_REQUEST_PATH = HttpData.ofUtf8(MSG_INVALID_REQUEST_PATH);
+    private final Consumer<ServerConfig> configUpdateListener = this::swapServerConfig;
```
Could you group this field with other member fields rather than static ones?

amitvc replied:
Yes.

trustin commented on core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java:
```diff
@@ -1567,14 +1575,15 @@ public Server build() {
             if (!Flags.useOpenSsl() && !SystemInfo.jettyAlpnOptionalOrAvailable()) {
                 throw new IllegalStateException(
                         "TLS configured but this is Java 8 and neither OpenSSL nor Jetty ALPN could be " +
-                        "detected. To use TLS with Armeria, you must either use Java 9+, enable OpenSSL, " +
-                        "usually by adding a build dependency on the " +
-                        "io.netty:netty-tcnative-boringssl-static artifact or enable Jetty ALPN as described " +
+                        "detected. To use TLS with Armeria, you must either use Java 9+, enable " +
+                        "OpenSSL, usually by adding a build dependency on the " +
+                        "io.netty:netty-tcnative-boringssl-static artifact or " +
+                        "enable Jetty ALPN as described " +
```
Could you explain why you needed this change?

amitvc replied:
If you are asking about this specific exception then I did not add this code. It is existing code. I can look at why this was added and report back to you.

minwoox replied:
any good news from here? >_<

trustin commented on core/src/main/java/com/linecorp/armeria/server/ServerConfig.java:
```diff
@@ -600,6 +606,14 @@ public boolean isDateHeaderEnabled() {
+     * Returns a map of SslContexts.
+     * @return Map of domain name to {@link SslContext}
+     */
+    public Mapping<String, SslContext> getSslContexts() {
```
- How about making this package-local so it's hidden from the public API?
- How about renaming to `sslContextMapping()`?

amitvc replied:
Good suggestions. Will make the change.

trustin commented on core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java:
```diff
@@ -154,7 +156,7 @@ static void safeClose(Channel ch) {
     }
 
-    private final ServerConfig config;
+    private volatile ServerConfigHolder configHolder;
```
Looks like this can be `final`, because `configHolder` is not replaced but only its content (`config`) is.

amitvc replied:
Yes. Will make the change

trustin commented on core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java:
```diff
@@ -313,9 +316,11 @@ private void handleRequest(ChannelHandlerContext ctx, DecodedHttpRequest req) th
         final RequestHeaders headers = req.headers();
         final String hostname = hostname(headers);
-        final VirtualHost virtualHost = config.findVirtualHost(hostname);
+        final VirtualHost virtualHost = configHolder.getConfig().findVirtualHost(hostname);
```
Could we call `configHolder.getConfig()` only once for each request, so that a request is handled with one single `ServerConfig` for consistency's sake? Let's declare a local variable and pass it as a parameter where necessary.

amitvc replied:
Makes sense. I will make the change.

trustin reviewed (commented):
Sorry that I'm late. Looks like we're going into the right direction. :+1:

amitvc commented:
@trustin @minwoox @ikhoon Thanks for the review. I will push changes tonight.

minwoox commented on core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java:
```diff
@@ -284,8 +289,8 @@ private void handleHttp2Settings(ChannelHandlerContext ctx, Http2Settings h2sett
         return new ServerHttp2ObjectEncoder(ctx, handler.encoder(), handler.keepAliveHandler(),
-                                            config.isDateHeaderEnabled(),
-                                            config.isServerHeaderEnabled()
+                                            configHolder.getConfig().isDateHeaderEnabled(),
```
Is there any reason that you used `configHolder.getConfig()` here?
We are using just `config` for other places in this class.

amitvc replied:
you are right. I think I missed this code. Will revert it back to config instead of configHolder.getConfig()

minwoox commented on core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java:
```diff
@@ -120,7 +121,8 @@
 
-        this.config = requireNonNull(config, "config");
+        configHolder = new ServerConfigHolder(requireNonNull(config, "config"));
+        this.config = configHolder.getConfig();
```
```suggestion
        this.config = config;
```

minwoox commented on core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java:
```diff
@@ -189,16 +191,22 @@ private void configureHttp(ChannelPipeline p, @Nullable ProxiedAddresses proxied
                 p.channel(), H1C, keepAliveHandler,
-                config.isDateHeaderEnabled(), config.isServerHeaderEnabled()
+                config.isDateHeaderEnabled(),
+                config.isServerHeaderEnabled()
```
revert?

minwoox commented on core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java:
```diff
@@ -189,16 +191,22 @@ private void configureHttp(ChannelPipeline p, @Nullable ProxiedAddresses proxied
+                                                                          responseEncoder,
+                                                                          H1C, proxiedAddresses);
+        p.addLast(httpServerHandler);
+        configHolder.addListener(httpServerHandler.configUpdateListener());
```
Shouldn't we do this for the `HttpServerHandler`s that are created in `Http2OrHttpHandler`?

amitvc replied:
Good catch. I have updated the code. Adding unit tests for https servers

minwoox commented on core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java:
```diff
@@ -455,13 +470,16 @@ private void addHttpHandlers(ChannelHandlerContext ctx) {
             final ServerHttp1ObjectEncoder writer = new ServerHttp1ObjectEncoder(
-                    ch, H1, keepAliveHandler, config.isDateHeaderEnabled(), config.isServerHeaderEnabled());
+                    ch, H1, keepAliveHandler, config.isDateHeaderEnabled(),
+                    config.isServerHeaderEnabled());
```
revert?

amitvc replied:
Fixed

minwoox commented on core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java:
```diff
@@ -560,7 +578,8 @@ private void configureHttp1WithUpgrade(ChannelHandlerContext ctx) {
 
-            addAfter(p, baseName, new Http1RequestDecoder(config, ctx.channel(), SCHEME_HTTP, responseEncoder));
+            addAfter(p, baseName, new Http1RequestDecoder(config,
+                                                          ctx.channel(), SCHEME_HTTP, responseEncoder));
```
revert?

amitvc replied:
Fixed

minwoox replied:
It's not fixed. >_<

minwoox commented on core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java:
```diff
@@ -1537,24 +1547,22 @@ public Server build() {
-                                   .map(vhb -> vhb.build(virtualHostTemplate))
-                                   .collect(toImmutableList());
-
+                        .map(vhb -> vhb.build(virtualHostTemplate))
+                        .collect(toImmutableList());
```
indentation and revert?

amitvc replied:
Fixed

minwoox commented on core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java:
```diff
@@ -1537,24 +1547,22 @@ public Server build() {
                 port -> checkState(port.protocols().stream().anyMatch(p -> p != PROXY),
-                                   "protocols: %s (expected: at least one %s or %s)",
-                                   port.protocols(), HTTP, HTTPS));
+                        "protocols: %s (expected: at least one %s or %s)",
+                        port.protocols(), HTTP, HTTPS));
```
ditto

amitvc replied:
Fixed.

minwoox commented on core/src/test/java/com/linecorp/armeria/client/TestReconfigurableServer.java:
```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2019 LINE Corporation
```
2021?

amitvc replied:
Changed

amitvc commented:
@minwoox thanks for the feedback on the PR. I will address the comments tonight.

amitvc commented:
@minwoox The PR is ready for review

minwoox commented on core/src/main/java/com/linecorp/armeria/server/Server.java:
```diff
@@ -397,6 +399,18 @@ public String toString() {
+    /**
+     * Reconfigure Server configuration.
+     */
+    public void reconfigure(ServerConfigurator rs) {
+        requireNonNull(rs, "serverconfigurator");
```
```suggestion
    public void reconfigure(ServerConfigurator serverConfigurator) {
        requireNonNull(serverConfigurator, "serverConfigurator");
```

minwoox commented on core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java:
```diff
@@ -189,16 +191,22 @@ private void configureHttp(ChannelPipeline p, @Nullable ProxiedAddresses proxied
     private Timer newKeepAliveTimer(SessionProtocol protocol) {
-        return MoreMeters.newTimer(config.meterRegistry(), "armeria.server.connections.lifespan",
+        return MoreMeters.newTimer(config.meterRegistry(),
+                                   "armeria.server.connections.lifespan",
```
Revert?
Could you revert all irrelevant changes, please? 😄

amitvc replied:
Reverted now. Sorry about the reformatting changes.

minwoox commented on core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java:
```diff
@@ -211,8 +219,9 @@ private void configureHttps(ChannelPipeline p, @Nullable ProxiedAddresses proxie
         final Timer keepAliveTimer = newKeepAliveTimer(scheme == SCHEME_HTTP ? H2C : H2);
-        return new Http2ServerConnectionHandlerBuilder(pipeline.channel(), config, keepAliveTimer,
-                                                       gracefulShutdownSupport, scheme.toString())
+        return new Http2ServerConnectionHandlerBuilder(pipeline.channel(), config,
```
Revert?

amitvc replied:
Done.

minwoox reviewed (commented):
Just nits from me. 😄

minwoox commented on core/src/test/java/com/linecorp/armeria/client/TestReconfigurableServer.java:
```diff
@@ -0,0 +1,225 @@
+        assertThat(response3.status()).isEqualTo(HttpStatus.NOT_FOUND);
+
+        final AggregatedHttpResponse response4 = client3.get("/test1/world")
+                .aggregate()
```
Could you also fix the code style of this class?
https://armeria.dev/community/developer-guide/#setting-up-your-ide

amitvc replied:
Done.

minwoox reviewed (commented):
Perhaps this is the final round from me.
Thanks a lot for your patience. 🙇‍♂️

amitvc commented on core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java:
```diff
@@ -692,4 +696,16 @@ private RequestId nextRequestId() {
+    void swapServerConfig(ServerConfig config) {
+        requireNonNull(config);
+        if (this.config != config) {
+            configHolder.replace(config);
```
Yes, you are correct. I think this is not necessary anymore.

amitvc commented on core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java:
```diff
@@ -193,8 +195,12 @@ private void configureHttp(ChannelPipeline p, @Nullable ProxiedAddresses proxied
+                                                                          responseEncoder,
+                                                                          H1C, proxiedAddresses);
+        p.addLast(httpServerHandler);
+        configHolder.addListener(httpServerHandler.configUpdateListener());
```
Done

ikhoon commented on core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java:
```diff
@@ -432,7 +442,10 @@ protected void configurePipeline(ChannelHandlerContext ctx, String protocol) thr
             final ChannelPipeline p = ctx.pipeline();
             p.addLast(newHttp2ConnectionHandler(p, SCHEME_HTTPS));
-            p.addLast(new HttpServerHandler(config, gracefulShutdownSupport, null, H2, proxiedAddresses));
+            final HttpServerHandler httpServerHandler = new HttpServerHandler(configHolder,
+                    gracefulShutdownSupport,
+                    null, H2, proxiedAddresses);
```
nit: I think the code style seems not to abide by our code formatter.
It should be either
```suggestion
            final HttpServerHandler httpServerHandler = new HttpServerHandler(configHolder,
                                                                              gracefulShutdownSupport,
                                                                              null, H2, proxiedAddresses);
```
or
```suggestion
            final HttpServerHandler httpServerHandler = new HttpServerHandler(
                    configHolder, gracefulShutdownSupport, null, H2, proxiedAddresses);
```

ikhoon commented on core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java:
```diff
@@ -461,7 +474,10 @@ private void addHttpHandlers(ChannelHandlerContext ctx) {
                     config.http1MaxChunkSize()));
             p.addLast(new Http1RequestDecoder(config, ch, SCHEME_HTTPS, writer));
-            p.addLast(new HttpServerHandler(config, gracefulShutdownSupport, writer, H1, proxiedAddresses));
+            final HttpServerHandler httpServerHandler = new HttpServerHandler(configHolder,
+                    gracefulShutdownSupport,
+                    writer, H1, proxiedAddresses);
```
nit: code style?

ikhoon commented on core/src/main/java/com/linecorp/armeria/server/Server.java:
```diff
@@ -397,6 +399,18 @@ public String toString() {
+        requireNonNull(pipelineConfigurator, "pipelineConfigurator");
+        final ServerBuilder sb = builder();
+        serverConfigurator.reconfigure(sb);
+        config = sb.buildServerConfig(config());
```
We need to call `ServiceCallbackInvoker.invokeServiceAdded()` if services are registered to a server. 
https://github.com/line/armeria/blob/4de44afc3017d7153e9bde67f8ff51e855877da1/core/src/main/java/com/linecorp/armeria/server/Server.java#L133-L135
Otherwise, the service will not work as expected.
https://github.com/line/armeria/blob/dbfc1cdbbacb7c15d8128b5b40972de28a82930e/grpc/src/main/java/com/linecorp/armeria/server/grpc/FramedGrpcService.java#L268-L272

trustin replied:
We also need to think about how we could notify the services being removed. Let me create a separate issue for it.

ikhoon commented on core/src/main/java/com/linecorp/armeria/server/Server.java:
```diff
@@ -397,6 +399,18 @@ public String toString() {
     }
 
+    /**
+     * Reconfigure Server configuration.
```
Could you add a description on
- `reconfigure()` should be called only after a `Server` has been started for the first time?
- Port newly added will not be applied.

amitvc replied:
Done

minwoox commented on core/src/main/java/com/linecorp/armeria/server/ServerConfigHolder.java:
```diff
@@ -0,0 +1,40 @@
+        this.config = config;
+    }
+
+    void replace(ServerConfig config) {
```
Because this method is not used anymore, I think we can remove this method and rename `replaceAndNotify(...)` to `replace(...)`

amitvc replied:
Done.

minwoox reviewed (approved):
🎉 🎉 🎉 
Thanks a lot, @amitvc!

trustin commented on core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java:
```diff
@@ -223,6 +227,8 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
                 // endOfStream set.
                 cleanup();
         }
+        //Cleanup the listener from ServerConfigHolder once the channel becomes inactive.
```
```suggestion
        // Clean up the listener from ServerConfigHolder once the channel becomes inactive.
```

amitvc replied:
Done.

trustin commented on core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java:
```diff
@@ -692,4 +697,15 @@ private RequestId nextRequestId() {
     }
+
+    void swapServerConfig(ServerConfig config) {
+        requireNonNull(config);
```
```suggestion
        requireNonNull(config, "config");
```

trustin commented on core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java:
```diff
@@ -692,4 +697,15 @@ private RequestId nextRequestId() {
+
+    void swapServerConfig(ServerConfig config) {
+        requireNonNull(config);
+        if (this.config != config) {
+            this.config = config;
+        }
```
No need to check identity when assigning:
```suggestion
        this.config = config;
```

trustin commented on core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java:
```diff
@@ -242,6 +247,11 @@ private Http2Settings http2Settings() {
     }
 
+    void updateConfig(ServerConfig config) {
+        final ServerConfig cfg = requireNonNull(config, "config");
+        configHolder.replaceAndNotify(cfg);
```
No need to introduce a local variable
```suggestion
        requireNonNull(config, "config");
        configHolder.replaceAndNotify(config);
```

amitvc replied:
Done.

trustin commented on core/src/test/java/com/linecorp/armeria/client/TestReconfigurableServer.java:
```diff
@@ -0,0 +1,206 @@
+class TestReconfigurableServer {
+
+    @Test
+    void test_reconfiguration_of_server() throws Exception {
```
Camel-case? How about: `shouldBeAbleToReconfigureServer()` ?

trustin reviewed (approved):
Almost there :+1:

amitvc commented:
@trustin @ikhoon @minwoox Thanks for the review. I will address the comments tonight and send another revision. Appreciate your patience and feedback

amitvc commented:
@trustin @minwoox I will address comments and send a new revision end of this week. I am traveling this week and have not been able to focus on this.

minwoox commented:
@amitvc No problem. Have a safe trip. 😄

amitvc commented on core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java:
```diff
@@ -461,7 +479,11 @@ private void addHttpHandlers(ChannelHandlerContext ctx) {
+                    gracefulShutdownSupport,
+                    writer, H1, proxiedAddresses);
+            p.addLast(httpServerHandler);
+            configHolder.addListener(httpServerHandler.configUpdateListener());
```
@ikhoon I have changed this already to match what you said here. 
[HttpServerPipelineConfigurator](https://github.com/amitvc/armeria/blob/3041_Dynamic_Reconfigure_Server/core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java#L477) 
[HttpServerHandler](https://github.com/amitvc/armeria/blob/3041_Dynamic_Reconfigure_Server/core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java#L194)

amitvc commented:
@minwoox @trustin @ikhoon I have addressed all comments.

ikhoon commented on core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java:
```diff
@@ -193,8 +195,11 @@ private void configureHttp(ChannelPipeline p, @Nullable ProxiedAddresses proxied
         p.addLast(new Http2PrefaceOrHttpHandler(responseEncoder));
-        p.addLast(new HttpServerHandler(config, gracefulShutdownSupport, responseEncoder,
-                                        H1C, proxiedAddresses));
+        final HttpServerHandler httpServerHandler = new HttpServerHandler(configHolder,
+                                                                          gracefulShutdownSupport,
+                                                                          responseEncoder,
+                                                                          H1C, proxiedAddresses);
+        p.addLast(httpServerHandler);
```
nit: Remove the local variable?

amitvc replied:
Done.

ikhoon commented on core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java:
```diff
@@ -432,7 +442,9 @@ protected void configurePipeline(ChannelHandlerContext ctx, String protocol) thr
             final ChannelPipeline p = ctx.pipeline();
             p.addLast(newHttp2ConnectionHandler(p, SCHEME_HTTPS));
-            p.addLast(new HttpServerHandler(config, gracefulShutdownSupport, null, H2, proxiedAddresses));
+            final HttpServerHandler httpServerHandler = new HttpServerHandler(configHolder,
+                    gracefulShutdownSupport, null, H2, proxiedAddresses);
+            p.addLast(httpServerHandler);
```
nit: Remove the local variable?

amitvc replied:
Yes. Done

ikhoon commented on core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java:
```diff
@@ -461,7 +473,9 @@ private void addHttpHandlers(ChannelHandlerContext ctx) {
                     config.http1MaxChunkSize()));
             p.addLast(new Http1RequestDecoder(config, ch, SCHEME_HTTPS, writer));
-            p.addLast(new HttpServerHandler(config, gracefulShutdownSupport, writer, H1, proxiedAddresses));
+            final HttpServerHandler httpServerHandler = new HttpServerHandler(configHolder,
+                    gracefulShutdownSupport, writer, H1, proxiedAddresses);
+            p.addLast(httpServerHandler);
```
nit: Remove the local variable?

amitvc replied:
Done

ikhoon commented on core/src/test/java/com/linecorp/armeria/client/TestReconfigurableServer.java:
```diff
@@ -0,0 +1,206 @@
+        assertThat(response11.contentUtf8()).isEqualTo("Hello, WORLD");
+
+        Thread.sleep(1000); // Some time goes by now want to reconfigure new server
+        System.out.println("Configuring new server");
```
Remove this?

amitvc replied:
Done

ikhoon commented on core/src/test/java/com/linecorp/armeria/client/TestReconfigurableServer.java:
```diff
@@ -0,0 +1,206 @@
+import com.linecorp.armeria.server.Server;
+import com.linecorp.armeria.server.ServerBuilder;
+
+class TestReconfigurableServer {
```
```suggestion
class ReconfigurableServerTest {
```

amitvc replied:
Done

ikhoon commented on core/src/test/java/com/linecorp/armeria/client/TestReconfigurableServer.java:
```diff
@@ -0,0 +1,206 @@
+        assertThat(res.status()).isEqualTo(HttpStatus.OK);
+        assertThat(res.contentUtf8()).isEqualTo("Hello, world!");
+
+        //Open new connection to check if the newly reconfigured server is now operational
```
```suggestion
        // Open new connection to check if the newly reconfigured server is now operational
```

amitvc replied:
Done

amitvc replied:
I ran checkstyl on my local and it passed. How do I catch these problems before I submit the CR?

ikhoon commented on core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java:
```diff
@@ -224,6 +228,8 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
                 // endOfStream set.
                 cleanup();
         }
+        //Clean up the listener from ServerConfigHolder once the channel becomes inactive.
```
```suggestion
        // Clean up the listener from ServerConfigHolder once the channel becomes inactive.
```

amitvc replied:
Done.

ikhoon commented on core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java:
```diff
@@ -193,8 +195,10 @@ private void configureHttp(ChannelPipeline p, @Nullable ProxiedAddresses proxied
         p.addLast(new Http2PrefaceOrHttpHandler(responseEncoder));
-        p.addLast(new HttpServerHandler(config, gracefulShutdownSupport, responseEncoder,
-                                        H1C, proxiedAddresses));
+        p.addLast(new HttpServerHandler(configHolder,
+                gracefulShutdownSupport,
+                responseEncoder,
+                H1C, proxiedAddresses));
```
nit: indent?

trustin replied:
```suggestion
        p.addLast(new HttpServerHandler(configHolder,
                                        gracefulShutdownSupport,
                                        responseEncoder,
                                        H1C, proxiedAddresses));
```

ikhoon commented on core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java:
```diff
@@ -432,7 +441,8 @@ protected void configurePipeline(ChannelHandlerContext ctx, String protocol) thr
             final ChannelPipeline p = ctx.pipeline();
             p.addLast(newHttp2ConnectionHandler(p, SCHEME_HTTPS));
-            p.addLast(new HttpServerHandler(config, gracefulShutdownSupport, null, H2, proxiedAddresses));
+            p.addLast(new HttpServerHandler(configHolder,
+                    gracefulShutdownSupport, null, H2, proxiedAddresses));
```
nit: indent?

amitvc replied:
I used intellij to reformat these lines of code. I have configured my ide to use Line corps settings.jar. Not sure what is the problem.

ikhoon commented on core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java:
```diff
@@ -461,7 +471,8 @@ private void addHttpHandlers(ChannelHandlerContext ctx) {
                     config.http1MaxChunkSize()));
             p.addLast(new Http1RequestDecoder(config, ch, SCHEME_HTTPS, writer));
-            p.addLast(new HttpServerHandler(config, gracefulShutdownSupport, writer, H1, proxiedAddresses));
+            p.addLast(new HttpServerHandler(configHolder,
+                    gracefulShutdownSupport, writer, H1, proxiedAddresses));
```
nit: indent?

amitvc replied:
I used intellij to reformat these lines of code. I have configured my ide to use Line corps settings.jar. Not sure what is the problem.

trustin commented on core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java:
```diff
@@ -432,7 +441,8 @@ protected void configurePipeline(ChannelHandlerContext ctx, String protocol) thr
             final ChannelPipeline p = ctx.pipeline();
             p.addLast(newHttp2ConnectionHandler(p, SCHEME_HTTPS));
-            p.addLast(new HttpServerHandler(config, gracefulShutdownSupport, null, H2, proxiedAddresses));
+            p.addLast(new HttpServerHandler(configHolder,
+                    gracefulShutdownSupport, null, H2, proxiedAddresses));
```
```suggestion
            p.addLast(new HttpServerHandler(configHolder,
                                            gracefulShutdownSupport
                                            null, H2, proxiedAddresses));
```

trustin commented on core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java:
```diff
@@ -461,7 +471,8 @@ private void addHttpHandlers(ChannelHandlerContext ctx) {
                     config.http1MaxChunkSize()));
             p.addLast(new Http1RequestDecoder(config, ch, SCHEME_HTTPS, writer));
-            p.addLast(new HttpServerHandler(config, gracefulShutdownSupport, writer, H1, proxiedAddresses));
+            p.addLast(new HttpServerHandler(configHolder,
+                    gracefulShutdownSupport, writer, H1, proxiedAddresses));
```
```suggestion
            p.addLast(new HttpServerHandler(configHolder,
                                            gracefulShutdownSupport,
                                            writer, H1, proxiedAddresses));
```

trustin replied:
Thanks @amitvc. I feel like we should really adopt an auto-formatter so we don't worry about this.

amitvc replied:
+1 on this. I typically select the code and use intelliji to reformat the code. Code -> Reformat code option on intellij.

trustin reviewed (approved):
<img src="https://media.giphy.com/media/6mYnVP9XUbRwXcaDiE/giphy.gif" width="160">

ikhoon reviewed (approved):
Nice work! Thanks for your endless efforts. 🙇‍♂️ @amitvc

amitvc commented:
@trustin @ikhoon @minwoox Thanks a lot for the reviews on this PR. I really appreciate our armeria community.

trustin commented:
> @trustin @ikhoon @minwoox Thanks a lot for the reviews on this PR. I really appreciate our armeria community.

I'm proud that you're part of it. Looking forward to your TLS trust manager PR! :heart:
